loopy/codegen/tools.py

```diff
@@ -0,0 +1,188 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from pytools import memoize_method
+from loopy.schedule import (EnterLoop, LeaveLoop, CallKernel, ReturnFromKernel,
+                            Barrier, BeginBlockItem, gather_schedule_block)
+
+
+__doc__ = """
+.. currentmodule:: loopy.codegen.tools
+
+.. autofunction:: make_codegen_cache_manager
+
+.. autoclass:: CodegenOperationCacheManager
+"""
+
+
+def make_codegen_cache_manager(kernel):
+    """
+    Returns an instance of :class:`CodegenOperationCacheManager` for *kernel*.
+    """
+    # {{{ computing active inames for sched indices
+
+    active_inames = []
+
+    for i in range(len(kernel.schedule)):
+        if i == 0:
+            active_inames.append(frozenset())
+            continue
+
+        sched_item = kernel.schedule[i-1]
+        if isinstance(sched_item, EnterLoop):
+            active_inames.append(active_inames[i-1]
+                    | frozenset([sched_item.iname]))
+        elif isinstance(sched_item, LeaveLoop):
+            assert sched_item.iname in active_inames[i-1]
+            active_inames.append(active_inames[i-1] - frozenset([sched_item.iname]))
+        else:
+            active_inames.append(active_inames[i-1])
+    # }}}
+
+    # {{{ computing call kernel indices
+
+    # callkernel_index[i]: schedule index of CallKernel containing the point
+    # just before ith-schedule item. None if the point is not a part of any
+    # callkernel.
+    callkernel_index = []
+
+    for i in range(len(kernel.schedule)):
+        if i == 0:
+            callkernel_index.append(None)
+            continue
+
+        sched_item = kernel.schedule[i-1]
+
+        if isinstance(sched_item, CallKernel):
+            callkernel_index.append(i-1)
+        elif isinstance(sched_item, ReturnFromKernel):
+            callkernel_index.append(None)
+        else:
+            callkernel_index.append(callkernel_index[i-1])
+
+    # }}}
+
+    # {{{ computing has_barrier_within
+
+    # has_barrier_within[i]: 'True' if the i-th schedule item's is a begin block
+    # containing a barrier of if the i-th schedule item is a barrier.
+
+    has_barrier_within = []
+
+    for sched_idx, sched_item in enumerate(kernel.schedule):
+        if isinstance(sched_item, BeginBlockItem):
+            # TODO: calls to "gather_schedule_block" can be amortized
+            _, endblock_index = gather_schedule_block(kernel.schedule,
+                    sched_idx)
+            has_barrier_within.append(any(
+                    isinstance(kernel.schedule[i], Barrier)
+                    for i in range(sched_idx+1, endblock_index)))
+        elif isinstance(sched_item, Barrier):
+            has_barrier_within.append(True)
+        else:
+            has_barrier_within.append(False)
+    # }}}
+
+    return CodegenOperationCacheManager(kernel, active_inames, callkernel_index,
+            has_barrier_within)
```

Comment:
How come these are computed ahead of time? (As opposed to on-demand, e.g. with `@memoize_method`?)

Reply:
With `memoize_method` we were running into problems of RecursionError. We also experiemented with getting away with it as: https://github.com/inducer/loopy/pull/179/commits/9c367039f96475d3a49f8bfdb9dbb9e502b7b93b. But, eventually agreed that computing all at once leads to the most maintainable code with a minor memory usage overhead.

Reply:
I don't understand what 9c36703 is supposed to do. I also don't understand why those three blocks of code can't just be packaged in memoized methods.

Reply:
For function `foo` if the `memoize_method` was called like `foo(0), foo(1), foo(2), ... , foo(10000)` in sequence then there'll be no recursion, but if you call `foo(10000)` at the beginning, then it will need to call `foo(9999)` which will call `foo(9998)` and so on leading to a recursion error.

Reply:
Sure, I understand. But what I'm suggesting is that these three entire computations (without any arguments) be moved into a memoized method.

Reply:
Done here: af108ef.